src/async/pitfalls/pin.md

```diff
@@ -0,0 +1,110 @@
+
+```rust,editable,compile_fail
+use tokio::sync::{mpsc, oneshot};
+use tokio::task::spawn;
+use tokio::time::{sleep, Duration};
+
+// A work item. In this case, just sleep for the given time and respond
+// with a message on the `respond_on` channel.
+#[derive(Debug)]
+struct Work {
+    input: u32,
+    respond_on: oneshot::Sender<u32>,
+}
+
+// A worker which listens for work on a queue and performs it.
+async fn worker(mut work_queue: mpsc::Receiver<Work>) {
+    let mut iterations = 0;
+    loop {
+        tokio::select! {
+            Some(work) = work_queue.recv() => {
+                sleep(Duration::from_millis(10)).await; // Pretend to work.
+                work.respond_on
+                    .send(work.input * 1000)
+                    .expect("failed to send response");
+                iterations += 1;
+            }
+            // TODO: report number of iterations every 100ms
+        }
+    }
+}
+
+// A requester which requests work and waits for it to complete.
+async fn do_work(work_queue: &mpsc::Sender<Work>, input: u32) -> u32 {
+    let (tx, rx) = oneshot::channel();
+    work_queue
+        .send(Work {
+            input,
+            respond_on: tx,
+        })
+        .await
+        .expect("failed to send on work queue");
+    rx.await.expect("failed waiting for response")
+}
+
+#[tokio::main]
+async fn main() {
+    let (tx, rx) = mpsc::channel(10);
+    spawn(worker(rx));
+    for i in 0..100 {
+        let resp = do_work(&tx, i).await;
+        println!("work result for iteration {i}: {resp}");
+    }
+}
+```
+
+<details>
+
+* You may recognize this as an example of the actor pattern. Actors
+  typically call `select!` in a loop.
+
+* This serves as a summation of a few of the previous lessons, so take your time
+  with it.
+
+    * Naively add a `_ = sleep(Duration::from_millis(100)) => { println!(..) }`
+      to the `select!`. This will never execute. Why?
+
+    * Instead, add a `timeout_fut` containing that future outside of the `loop`:
+
+        ```rust,compile_fail
+        let mut timeout_fut = sleep(Duration::from_millis(100));
+        loop {
+            select! {
+                ..,
+                _ = timeout_fut => { println!(..); },
+            }
+        }
+        ```
+    * This still doesn't work. Follow the compiler errors, adding `&mut` to the
+      `timeout_fut` in the `select!` to work around the move, then using
+      `Box::pin`:
+
+        ```rust,compile_fail
+        let mut timeout_fut = Box::pin(sleep(Duration::from_millis(100)));
+        loop {
+            select! {
+                ..,
+                _ = &mut timeout_fut => { println!(..); },
+            }
+        }
+        ```
+
+    * This compiles, but once the timeout expires it is `Poll::Ready` on every
+      iteration (a fused future would help with this). Update to reset
+      `timeout_fut` every time it expires.
+
+* Box allocates on the heap. In some cases, `tokio::pin!` is also an option, but
```

Comment:
Maybe prefer or mention `std::pin::pin!`?

We're at an awkward point where the standard library has a pin macro (as of 1.68, stable on 2023-03-09) but
* It might not be supported if they haven't updated, or used Rust from a distro
* Folks are likely to run into the Tokio macro in existing code